openedx_events/event_bus/avro/tests/test_avro.py

```diff
@@ -11,10 +17,66 @@
     deserialize_bytes_to_event_data,
     serialize_event_data_to_bytes,
 )
+from openedx_events.learning import signals  # pylint: disable=reimported
+from openedx_events.tests.utils import FreezeSignalCacheMixin
+from openedx_events.tooling import OpenEdxPublicSignal
+
+# when adding new signals, note which signals are not for use with the event bus by
+# adding it to the known unserializable signals dictionary or fix the signal
+KNOWN_UNSERIALIZABLE_SIGNALS = ["org.openedx.learning.discussions.configuration.changed.v1"]
+
+# TODO (EventBus): figure out how to automatically pull new signals in
+
+
+def generate_test_event_data_for_data_type(data_type):
+    """
+    Generates test data automatically by filling in dummy data for basic data types (int/bool/str)
+    Recursively breaks down the classes for nested classes
+
+    Arguments:
+        data_type: The type of the data which we are generating data for
+
+    Returns:
+        dtdict: A data dictionary containing dummy data for all attributes of the class
+    """
+    dtdict = {}
+    defaults_per_type = {
+        int: 1,
+        bool: True,
+        str: "default",
+        CourseKey: CourseKey.from_string("course-v1:edX+DemoX.1+2014"),
+        datetime: datetime.now(),
+    }
+    for attribute in data_type.__attrs_attrs__:
+        result = defaults_per_type.get(attribute.type, None)
+        if result is not None:
+            dtdict.update({attribute.name: result})
+        elif attribute.type in [dict, list]:
+            raise Exception("Unable to generate Avro schema for dict or array fields")
+        else:
+            dtdict.update({attribute.name: attribute.type(**generate_test_event_data_for_data_type(attribute.type))})
+    return dtdict
 
 
-class TestAvro(TestCase):
+class TestAvro(FreezeSignalCacheMixin, TestCase):
     """Tests for end-to-end serialization and deserialization of events"""
+    def test_all_events(self):
+        for signal in OpenEdxPublicSignal.all_events():
+            if signal.event_type not in KNOWN_UNSERIALIZABLE_SIGNALS:
+                test_data = {}
+                serializer = AvroSignalSerializer(signal)
+                attribute_number = 0
+                for b in list(signal.init_data.keys()):
```

Comment:
I think I misled you here. I forgot that `schema.fields` is a list and not a dict. We cannot guarantee that the order of the fields in `schema.fields` is the same as the order of the keys we get in `signal.init_data.keys()` . 

Anyway, it makes more sense to generate the test data based solely on what's in the signal  (not the schema), since the schema is part of what we're testing. 

In light of that, we should just be able to iterate through everything in the `signal.init_data` dict and use the key for the attribute name and the value as the current_class. this will avoid having to iterate with list indexes and will not depend on our particular implementation of the schema generator.